File: src/main/scala/org/cyclops/fluidconverters/FluidColorAnalyzer.scala
package org.cyclops.fluidconverters

import java.awt.image.BufferedImage
import java.io.FileNotFoundException
import javax.imageio.ImageIO

import net.minecraft.client.Minecraft
import net.minecraft.util.{ResourceLocation, IIcon}
import net.minecraftforge.fluids.{FluidRegistry, Fluid}
import org.cyclops.fluidconverters.config.{FluidGroup, FluidGroupRegistry}


/**
 * Analyze the colors of fluids and fluid groups.
 * @author rubensworks
 */
object FluidColorAnalyzer {

    val colors = scala.collection.mutable.Map[Fluid, Int]()
    val groupColors = scala.collection.mutable.Map[FluidGroup, Int]()
    final val DEFAULT = 16777215

    def init() {
        val it = FluidRegistry.getRegisteredFluids.values().iterator()
        while(it.hasNext) {
            val fluid = it.next()
            colors.put(fluid, calculateAverageColor(fluid))
        }

        val it2 = FluidGroupRegistry.getGroups.iterator
        while(it2.hasNext) {
            val group = it2.next()
            groupColors.put(group, calculateAverageColor(group))
        }
    }

    def getAverageColor(fluidGroup: FluidGroup): Int = {
        groupColors.get(fluidGroup) match {
            case Some(i) => i
            case None => DEFAULT
        }
    }

    private def calculateAverageColor(fluidGroup: FluidGroup): Int = {
        var r = 0F
        var g = 0F
        var b = 0F
        for(element <- fluidGroup.getFluidElements) {
            val fluid = element.getFluid
            val color = getAverageColor(fluid)
            r += (-color >> 16 & 0xFF) / 255F
            g += (-color >> 8 & 0xFF) / 255F
            b += (-color & 0xFF) / 255F
        }
        val l = fluidGroup.getFluidElements.length
        r /= l
        g /= l
        b /= l

        (r * 255).toInt << 16 | (g * 255).toInt << 8 | (b * 255).toInt
    }

    def getAverageColor(fluid: Fluid): Int = {
        colors.get(fluid) match {
            case Some(i) => i
            case None => DEFAULT
        }
    }

    private def calculateAverageColor(fluid: Fluid): Int = {
        val default = fluid.getColor

        val block = fluid.getBlock
        if(block == null) {
            return default
        }

        val icon = block.getIcon(0, 0)
        if(icon == null) {
            return default
        }

        val image = readIcon(icon)
        if(image == null) {
            return default
        }

        val width = image.getWidth
        val height = image.getHeight
        val matrix = new Array[Int](width * height)
        image.getRGB(0, 0, width, height, matrix, 0, width)

        if(matrix.length == 0) {
            return default
        }

        var r = 0F
        var g = 0F
        var b = 0F
        matrix.foreach((i) => {
            r += (-i >> 16 & 0xFF) / 255F
            g += (-i >> 8 & 0xFF) / 255F
            b += (-i & 0xFF) / 255F
        })
        r /= matrix.length
        g /= matrix.length
        b /= matrix.length

        (r * 255).toInt << 16 | (g * 255).toInt << 8 | (b * 255).toInt
    }

    private def readIcon(icon: IIcon): BufferedImage = {
        // Determine ResourceLocation params
        val iconName = icon.getIconName
        var iconDomain = "minecraft"
        var iconPath = iconName
        val index = iconPath.indexOf(':')
        if(index >= 0) {
            iconPath = iconName.substring(index + 1, iconName.length)
            if(index > 1) {
                iconDomain = iconName.substring(0, index)
            }
        }
        // Read the resource location
        try {
            val resourceLocation = new ResourceLocation(iconDomain.toLowerCase, "textures/blocks/%s.png".format(iconPath))
            val resource = Minecraft.getMinecraft.getResourceManager.getResource(resourceLocation)
            ImageIO.read(resource.getInputStream)
        } catch {
            case e: FileNotFoundException => return null
        }
    }

}

File: src/main/java/org/cyclops/fluidconverters/FluidColorAnalyzer.java
package org.cyclops.fluidconverters;

import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.fluids.Fluid;
import org.apache.commons.lang3.tuple.Triple;
import org.cyclops.cyclopscore.helper.Helpers;
import org.cyclops.cyclopscore.helper.RenderHelpers;
import org.cyclops.fluidconverters.fluidgroup.FluidGroup;
import org.cyclops.fluidconverters.fluidgroup.FluidGroupRegistry;
import org.lwjgl.util.Color;

import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/**
 * Analyzes the colors of fluids in fluid groups and calculates average colors.
 * @author immortaleeb
 */
public class FluidColorAnalyzer {

    private static final Map<FluidGroup, Color> fluidGroupColors = new HashMap<FluidGroup, Color>();

    /**
     * Forces the color analyzer to calculate all average colors for all fluid
     * groups and cache them for later use.
     */
    public static void calculateAverageColors() {
        Iterator<FluidGroup> it = FluidGroupRegistry.iterator();
        while (it.hasNext()) {
            FluidGroup group = it.next();
            fluidGroupColors.put(group, calculateAverageColor(group));
        }
    }

    /**
     * Gets the cached average color of the given fluid group
     * @param fluidGroup the fluid group
     * @return The average color of all fluids in the fluid group
     */
    public static Color getAverageColor(FluidGroup fluidGroup) {
        return fluidGroupColors.get(fluidGroup);
    }

    /**
     * Calculates the average color of a single fluid
     * @param fluid The fluid for which we need to calculate the average color
     * @return Average color of the fluid, or white in case it could not be calculated
     */
    private static Color calculateAverageColor(Fluid fluid) {
        TextureAtlasSprite icon = RenderHelpers.getFluidIcon(fluid, EnumFacing.UP);
        int[] texture = icon.getFrameTextureData(0)[0];
        float r = 0, g = 0, b = 0;
        for (int pixel : texture) {
            Triple<Float, Float, Float> triple = Helpers.intToRGB(pixel);
            r += triple.getLeft();
            g += triple.getMiddle();
            b += triple.getRight();
        }

        r /= texture.length;
        g /= texture.length;
        b /= texture.length;

        return new Color((int) (r * 255), (int) (g * 255), (int) (b * 255));
    }

    /**
     * Calculates the average color of all fluids in the fluidgroup
     * @param fluidGroup the fluid group
     * @return The average color of all fluids in the fluid group, or white in case it could not be calculated
     */
    private static Color calculateAverageColor(FluidGroup fluidGroup) {
        int r = 0;
        int g = 0;
        int b = 0;

        List<FluidGroup.FluidElement> fluidElements = fluidGroup.getFluidElements();
        int size = fluidElements.size();
        for (FluidGroup.FluidElement fluidElement : fluidElements) {
            Color color = calculateAverageColor(fluidElement.getFluid());
            r += color.getRed();
            g += color.getGreen();
            b += color.getBlue();
        }

        r = r / size;
        g = g / size;
        b = b / size;

        return new Color(r, g, b);
    }
}
